src/Microsoft.VisualStudio.Services.Agent/Condition.tokens.cs

```diff
@@ -0,0 +1,350 @@
+                str.Equals(Constants.Conditions.NotEqual, StringComparison.OrdinalIgnoreCase) ||
+                str.Equals(Constants.Conditions.Or, StringComparison.OrdinalIgnoreCase) ||
+                str.Equals(Constants.Conditions.Xor, StringComparison.OrdinalIgnoreCase))
+            {
+                _trace.Verbose($"Fnctn: {str}");
+                _tokens.Add(new FunctionToken(str, startIndex, length));
+            }
+            else if (str.Equals(Constants.Conditions.Capabilities, StringComparison.OrdinalIgnoreCase) ||
+                str.Equals(Constants.Conditions.Variables, StringComparison.OrdinalIgnoreCase))
+            {
+                _trace.Verbose($"Hasht: {str}");
+                _tokens.Add(new HashtableToken(str, startIndex, length));
+            }
+            else
+            {
+                _trace.Verbose($"Unrec: {str}");
+                _tokens.Add(new UnrecognizedToken(startIndex, length));
+            }
+        }
+
+        private void CreateStringToken(ref int index)
+        {
+            // TODO: Confirm double-single-quote for escaping is sufficient. Better than backslash-escaping since this is not a complex language and backslash is common to file-paths.
+            int startIndex = index;
+            char c;
+            bool closed = false;
+            var str = new StringBuilder();
+            index++; // Skip the leading single-quote.
+            while (index < _raw.Length)
+            {
+                c = _raw[index++];
+                if (c == '\'')
+                {
+                    // End of string.
+                    if (index >= _raw.Length || _raw[index] != '\'')
+                    {
+                        closed = true;
+                        break;
+                    }
+
+                    // Escaped single quote.
+                    index++;
+                }
+
+                str.Append(c);
+            }
+
+            int length = index - startIndex;
+            if (closed)
+            {
+                _trace.Verbose($"Strng: '{str.ToString()}'");
+                _tokens.Add(new StringToken(str.ToString(), startIndex, length));
+                return;
+            }
+
+            _trace.Verbose($"InvSt: '{str.ToString()}'");
+            _tokens.Add(new UnterminatedStringToken(startIndex, length));
+        }
+
+        private static bool IsWhitespaceOrPunctuation(char c)
+        {
+            switch (c)
+            {
+                case Constants.Conditions.CloseFunction:
+                case Constants.Conditions.CloseHashtable:
+                case Constants.Conditions.OpenFunction:
+                case Constants.Conditions.OpenHashtable:
+                case Constants.Conditions.Separator:
+                    return true;
+                default:
+                    return char.IsWhiteSpace(c);
+            }
+        }
+
+        private abstract class Token
```

Comment:
```
token
{
  enum kind: , ( ) [ ] number string bool and or variables capabilities unrecognized (dont need malformed-number or unterminated-string)
  string value
}
```

Punctuation: `, ( ) [ ]`
Literal values: number, string, bool
Keywords: and or variables